### PR TITLE
Fix reporting of EnC (and other) diagnostics

### PR DESCRIPTION
**Pull request by tmat**
Fixes issues with reporting diagnostics from EnC. Rude edits were sometimes not reported, EnC delta emit errors that didn't have a location were never reported.

There were a couple of bugs in the code base:
1) Logic that grouped diagnostics by document dropped ones that have no document (project-level)
2) Compiler API ```CompilationWithAnalyzers.GetEffectiveDiagnostics``` was implemented as an iterator and incorrectly used in analyzer executor. The usage pattern was:

```C#
using (var pooledList = GetListFromPool())
{
    Populate(pooledList);
    return CompilationWithAnalyzers.GetEffectiveDiagnostics(pooledList);
}
```

The caller returned before pulling on the iterator. As a result the enumerator was actually iterating over a released ```pooledList``` instance.

This change refactors the code to use ImmutableArray instead of IEnumerable and changes the public API to eagerly read the input IEnumerable (it seems that that was the original intent). This is slightly breaking change to the API (from lazy to eager) but it's imo more likely to fix lingering issues
than to actually break code.

Fixes https://github.com/dotnet/roslyn/issues/11152, https://github.com/dotnet/roslyn/issues/10040, https://github.com/dotnet/roslyn/issues/3634.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-analysis Please review changes in the analyzer executor.
@dotnet/roslyn-compiler @gafter @AnthonyDGreen Please review the compiler public API change.
@dotnet/roslyn-interactive  Please review the EnC changes.

**Comment by AlekseyTs** [comment]
Changes under src/Compilers LGTM

**Comment by jaredpar** [comment]
test eta

**Comment by jaredpar** [comment]
test vsi

**Comment by jaredpar** [comment]
test windows_vsi_p0_open_prtest please

**Comment by jaredpar** [comment]
test vsi please

**Comment by jaredpar** [comment]
@tmat you can ignore the VSI / ETA tests.  I needed a PR that wasn't mine to test the Jenkins changes I've been making.  Thanks for helping :wink:

File: src/EditorFeatures/Test/Workspaces/NoCompilationDocumentDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -17,16 +18,15 @@ internal class NoCompilationDocumentDiagnosticAnalyzer : DocumentDiagnosticAnaly
 
         public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
 
-        public override Task AnalyzeSemanticsAsync(Document document, Action<Diagnostic> addDiagnostic, CancellationToken cancellationToken)
+        public override Task<ImmutableArray<Diagnostic>> AnalyzeSemanticsAsync(Document document, CancellationToken cancellationToken)
```
Comment (heejaechang) [inline src/EditorFeatures/Test/Workspaces/NoCompilationDocumentDiagnosticAnalyzer.cs]: this is breaking change for F# and typescript, I am find doing this way but you should make sure typescript and F# to change as well.

**Comment by heejaechang** [comment]
@mavasani want to take a look? I guess we will remove v1 code soon?

**Comment by balajikris** [comment]
vsi p1 failure is https://github.com/dotnet/roslyn/issues/10416
vsi p0 failure is https://github.com/dotnet/roslyn/issues/9602

**Comment by tmat** [comment]
@heejaechang I have added the original overloads back for now until the changes to TS and F# are made.

**Comment by tmat** [comment]
test vsi

**Comment by tmat** [comment]
test vsi please

File: src/EditorFeatures/Test/Workspaces/NoCompilationDocumentDiagnosticAnalyzer.cs (hunk 1 above)
Comment (mavasani) [inline src/EditorFeatures/Test/Workspaces/NoCompilationDocumentDiagnosticAnalyzer.cs]: Yes, TS will need to get fixed up with insertion.

**Comment by mavasani** [comment]
@heejaechang - Yes we should remove v1 engine post Update3. Filed https://github.com/dotnet/roslyn/issues/11464

**Comment by mavasani** [comment]
👍

**Comment by tmat** [comment]
test vsi please

**Pull request merged**